### ISSUE TITLE
[regression] Error computing .def files location when using subprojects

### ISSUE THREAD

**Issue by thiblahute** [mesonbuild/meson#722]
Since:

```
commit 88037c43858c600d5004ede1ab01378cb110c8af
Author: Nirbheek Chauhan <nirbheek@centricular.com>
Date:   Thu Aug 18 20:52:22 2016 +0530

    shared_library: Add vs_module_defs to link_depends

    With this, if the module definitions file is edited, the shared library
    and all reverse-dependencies will be re-linked.

    Fixes #643

:040000 040000 84e77a840d469fc4e4821733e18542e92f0dfaa5 c7ea8aca66e8e4150e23966e75ae525375f8f330 M  mesonbuild
```

I am getting the following error:

```
ninja: Entering directory `mesonbuild'
ninja: error: '/home/thiblahute/devel/gstreamer/master/gst-all/win32/common/libgstreamer.def', needed by 'subprojects/gstreamer/gst/libgstreamer-1.0.so.0.901.0', missing and no known rule to make it
ninja: Entering directory `mesonbuild'
ninja: error: '/home/thiblahute/devel/gstreamer/master/gst-all/win32/common/libgstreamer.def', needed by 'subprojects/gstreamer/gst/libgstreamer-1.0.so.0.901.0', missing and no known rule to make it
```

Trying to build [gst-all](https://github.com/thiblahute/gst-all)

**Comment by tp-m** [comment]
Is this the issue we talked about the other day, that source_root for the subproject expands to the source root of the parent project instead of the subproject source root?

**Comment by thiblahute** [comment]
I might be related indeed.

**Comment by jpakkane** [comment]
The core of this issue seems to be whether `meson.source_root()` resolve to the toplevel root directory or the top level of the current subproject. If we want to keep the semantic meaning of subprojects, roughly behave-as-a-chroot, then the latter makes sense.

The question then becomes whether subprojects ever need to know where the topmost source dir is. I can't think of a use case for this and if there is one we can add a method like `meson.get_global_source_root()` or equivalent.

**Comment by tp-m** [comment]
Agree that the chroot-like/isolated behaviour makes the most sense, and that we can always add something else in case a subproject really needs to know about a superproject.

**Comment by tp-m** [comment]
Agree that the chroot-like/isolated behaviour makes the most sense, and that we can always add something else in case a subproject really needs to know about a superproject.

**Comment by tp-m** [comment]
Agree that the chroot-like/isolated behaviour makes the most sense, and that we can always add something else in case a subproject really needs to know about a superproject.

**Comment by eli-schwartz** [comment]
#761 got closed as "the wrong approach", but .project_*_root() methods got added in #7772 recently, meaning this is fixed in 0.56
